File: HFNavigationController/HFNavigationController.m
//
//  HFNavigationController.m
//  HFNavigationController
//
//  Created by xhf on 16/5/11.
//  Copyright © 2016年 xuhongfei. All rights reserved.
//

#import "HFNavigationController.h"
#import "AppDelegate.h"


#define HF_KEY_WINDOW  ((AppDelegate *)[UIApplication sharedApplication].delegate).window
#define HF_TOP_VIEW  ((AppDelegate *)[UIApplication sharedApplication].delegate).window.rootViewController.view
#define HF_SCREEN_WIDTH [UIScreen mainScreen].bounds.size.width

@interface HFNavigationController () <UIGestureRecognizerDelegate>
{
    CGPoint startTouch;
    UIImageView *lastScreenShotView;
    UIView *blackMask;
}

@property (nonatomic, strong) UIView *backgroundView;
@property (nonatomic, strong) NSMutableArray *screenShotsList;
@property (nonatomic, assign) BOOL isMoving;

@end

@implementation HFNavigationController

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        self.screenShotsList = [NSMutableArray arrayWithCapacity:0];
        self.canInteractive = YES;
        self.popAnimationType = PopAnimationTypeLiner;
        self.navigationBar.translucent = false;
        
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIScreenEdgePanGestureRecognizer *screenEdgePanGP = [[UIScreenEdgePanGestureRecognizer alloc] initWithTarget:self action:@selector(panGestureRecognizer:)];
    screenEdgePanGP.edges = UIRectEdgeLeft;
    screenEdgePanGP.delegate = self;
    [self.view addGestureRecognizer:screenEdgePanGP];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (self.screenShotsList.count == 0) {
        UIImage *capturedImage = [self capture];
        if (capturedImage) {
            [self.screenShotsList addObject:capturedImage];
        }
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)dealloc
{
    self.screenShotsList = nil;
    self.backgroundView = nil;
}

#pragma mark - Get Set
- (void)setPopAnimationType:(PopAnimationType)popAnimationType
{
    _popAnimationType = popAnimationType;
}

#pragma mark -
- (void)pushViewController:(UIViewController *)viewController animated:(BOOL)animated
{
    UIImage *capturedImage = [self capture];
    if (capturedImage) {
        [self.screenShotsList addObject:capturedImage];
    }
    
    if (self.viewControllers.count) {
        viewController.navigationItem.leftBarButtonItem = [self itemWithImage:@"tab_return" higlightedImage:@"tab_return" target:self action:@selector(back)];
    }
    
    if (self.viewControllers.count == 1) {
        viewController.hidesBottomBarWhenPushed = YES;
    }
    
    [super pushViewController:viewController animated:animated];
}

- (void)back
{
    [self popViewControllerAnimated:YES];
}

- (UIBarButtonItem *)itemWithImage:(NSString *)image higlightedImage:(NSString *)higlightedImage  target:(id)target action:(SEL)action
{
    // 1.创建按钮
    UIButton *btn = [UIButton buttonWithType:UIButtonTypeCustom];
    
    // 2.设置按钮背景图片
    UIImage *normal = [UIImage imageNamed:image];
    [btn setBackgroundImage:normal forState:UIControlStateNormal];
    [btn setBackgroundImage:[UIImage imageNamed:higlightedImage] forState:UIControlStateHighlighted];
    
    // 3.设置按钮的尺寸
    btn.bounds = CGRectMake(0, 0, 24, 24);
    
    // 4.监听按钮点击
    [btn addTarget:target action:action forControlEvents:UIControlEventTouchUpInside];
    
    // 5.返回创建好的item
    return [[UIBarButtonItem alloc] initWithCustomView:btn];
}

- (UIViewController *)popViewControllerAnimated:(BOOL)animated
{
    [self.screenShotsList removeLastObject];
    
    return [super popViewControllerAnimated: animated];
}

#pragma mark -
- (UIImage *)capture
{
    UIGraphicsBeginImageContextWithOptions(HF_TOP_VIEW.bounds.size, HF_TOP_VIEW.opaque, 0.0);
    [HF_TOP_VIEW.layer renderInContext:UIGraphicsGetCurrentContext()];
    
    UIImage * img = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return img;
}


#pragma mark -
- (void)panGestureRecognizer:(UIScreenEdgePanGestureRecognizer *)recoginzer
{
    // If the viewControllers has only one vc or disable the interaction, then return.
    if (self.viewControllers.count <= 1 || !self.canInteractive) return;
    
    // we get the touch position by the window's coordinate
    CGPoint touchPoint = [recoginzer locationInView:HF_KEY_WINDOW];
    
    // begin paning, show the backgroundView(last screenshot),if not exist, create it.
    if (recoginzer.state == UIGestureRecognizerStateBegan) {
        
        _isMoving = YES;
        startTouch = touchPoint;
        
        if (!self.backgroundView)
        {
            CGRect frame = HF_TOP_VIEW.frame;
            self.backgroundView = [[UIView alloc]init];
            CGRect backgroundViewFrame = CGRectMake(0, 0, frame.size.width , frame.size.height);
            self.backgroundView.frame = backgroundViewFrame;
            self.backgroundView.backgroundColor = [UIColor whiteColor];
            [HF_TOP_VIEW.superview insertSubview:self.backgroundView belowSubview:HF_TOP_VIEW];
            
            blackMask = [[UIView alloc]initWithFrame:CGRectMake(0, 0, frame.size.width , frame.size.height)];
            blackMask.backgroundColor = [UIColor blackColor];
            [self.backgroundView addSubview:blackMask];
        }
        
        self.backgroundView.hidden = NO;
        
        if (lastScreenShotView) [lastScreenShotView removeFromSuperview];
        
        UIImage *lastScreenShot = [self.screenShotsList lastObject];
        lastScreenShotView = [[UIImageView alloc]initWithImage:lastScreenShot];
        [self.backgroundView insertSubview:lastScreenShotView belowSubview:blackMask];
        
        //End paning, always check that if it should move right or move left automatically
    }else if (recoginzer.state == UIGestureRecognizerStateEnded){
        
        if (touchPoint.x - startTouch.x > HF_SCREEN_WIDTH * .4f)
        {
            [UIView animateWithDuration:0.3 animations:^{
                [self moveViewWithX: HF_SCREEN_WIDTH];
            } completion:^(BOOL finished) {
                
                [self popViewControllerAnimated:NO];
                CGRect frame = HF_TOP_VIEW.frame;
                frame.origin.x = 0;
                HF_TOP_VIEW.frame = frame;
                
                _isMoving = NO;
                self.backgroundView.hidden = YES;
                
            }];
        }
        else
        {
            [UIView animateWithDuration:0.3 animations:^{
                [self moveViewWithX:0];
            } completion:^(BOOL finished) {
                _isMoving = NO;
                self.backgroundView.hidden = YES;
            }];
        }
        return;
        
        // cancal panning, alway move to left side automatically
    }else if (recoginzer.state == UIGestureRecognizerStateCancelled){
        
        [UIView animateWithDuration:0.3 animations:^{
            [self moveViewWithX:0];
        } completion:^(BOOL finished) {
            _isMoving = NO;
            self.backgroundView.hidden = YES;
        }];
        
        return;
    }
    
    // it keeps move with touch
    if (_isMoving) {
        [self moveViewWithX:touchPoint.x - startTouch.x];
    }
}

// set lastScreenShotView 's position and alpha when paning
- (void)moveViewWithX:(float)x
{
    NSLog(@"Move to:%f",x);
    x = x > HF_SCREEN_WIDTH ? HF_SCREEN_WIDTH : x;
    x = x < 0 ? 0 : x;
    
    CGRect frame = HF_TOP_VIEW.frame;
    frame.origin.x = x;
    HF_TOP_VIEW.frame = frame;
    
    float alpha = 0.5 - (x / HF_SCREEN_WIDTH * .5f);
    float scale = (x / HF_SCREEN_WIDTH * 0.05) + 0.95;

    switch (self.popAnimationType) {
        case PopAnimationTypeLiner:
            lastScreenShotView.transform = CGAffineTransformMakeTranslation( x / 2.f - lastScreenShotView.center.x, 0);
            break;
        case PopAnimationTypeScale:
            lastScreenShotView.transform = CGAffineTransformMakeScale(scale, scale);
            break;
        default:
            break;
    }
    

    blackMask.alpha = alpha;
    
}

#pragma mark - UIGestureRecognizerDelegate
- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
{
    return self.viewControllers.count > 1 && self.canInteractive ? YES : NO;
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: HFNavigationController/HFNavigationController/swift/HFNavigationController.swift
//
//  HFNavigationController.swift
//  DcoinWallet
//
//  Created by hongfei xu on 2018/11/6.
//  Copyright © 2018 dcoin. All rights reserved.
//

import UIKit

class HFNavigationController: UINavigationController {

    // MARK: - public
    
    public enum PopAnimationType {
        /// 线性动画，default
        case liner
        /// 幕布式动画
        case curtain
        /// 缩放式动画
        case scale
    }
    
    /// 是否支持手势返回
    public var canIneractive: Bool = true
    
    /// pop动画类型
    public var popAnimationType: PopAnimationType = .liner
    
    // MARK: - private
    fileprivate var screenShots: [UIImage] = []
    fileprivate var isMoving: Bool = false
    fileprivate var startTouch: CGPoint = CGPoint.zero
    
    fileprivate var backgroundView: UIView?
    fileprivate var blackMask: UIView?
    fileprivate var lastScreenShotView: UIImageView?
    
    fileprivate var keyWindow: UIWindow {
        get {
            if let delegate = UIApplication.shared.delegate as? AppDelegate {
                return delegate.window!
            }
            return UIApplication.shared.keyWindow!
        }
    }
    
    fileprivate var topView: UIView? {
        get {
            return keyWindow.rootViewController?.view
//            return UIApplication.shared.keyWindow?.rootViewController?.view
        }
    }
    
    fileprivate var screenWidth: CGFloat {
        get {
            return UIScreen.main.bounds.size.width
        }
    }
    
    // MARK: - life cycle
    override func viewDidLoad() {
        super.viewDidLoad()

        setup()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if self.screenShots.count == 0 {
            if let capturedImage = capture() {
                screenShots.append(capturedImage)
            }
        }
    }
    
    deinit {
        screenShots = []
    }

    // MARK: -
    fileprivate func setup() {
        
        navigationBar.isTranslucent = false
        
        let screenEdgePanGR = UIScreenEdgePanGestureRecognizer(target: self, action: #selector(panGestureRecognizer(_:)))
        screenEdgePanGR.edges = .left
        screenEdgePanGR.delegate = self
        view.addGestureRecognizer(screenEdgePanGR)
        
    }
    
    // MARK: -
    override func pushViewController(_ viewController: UIViewController, animated: Bool) {
        if let capturedImage = capture() {
            screenShots.append(capturedImage)
        }
        
        if viewControllers.count > 0 {
    
            viewController.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "back", style: .plain, target: self, action: #selector(back))
        }
        
        if viewControllers.count == 1 {
            viewController.hidesBottomBarWhenPushed = true
        }
        
        super.pushViewController(viewController, animated: animated)
    }
    
    override func popViewController(animated: Bool) -> UIViewController? {
        
        screenShots.removeLast()
        
        return super.popViewController(animated: animated)
    }
    
    // MARK: -
    @objc
    fileprivate func panGestureRecognizer(_ panGR: UIScreenEdgePanGestureRecognizer) {
        
        guard viewControllers.count > 1 && canIneractive else {
            return
        }
        
        let touchPoint = panGR.location(in: keyWindow)
        
        switch panGR.state {
        case .began:
            
            isMoving = true
            startTouch = touchPoint
            
            if backgroundView == nil {
                let frame = topView!.frame
                backgroundView = UIView()
                let bgvFrame = CGRect(origin: CGPoint.zero, size: frame.size)
                backgroundView!.frame = bgvFrame
                backgroundView!.backgroundColor = UIColor.white
                topView!.superview?.insertSubview(backgroundView!, belowSubview: topView!)
                
                blackMask = UIView(frame: CGRect(origin: CGPoint.zero, size: frame.size))
                blackMask?.backgroundColor = UIColor.black
                backgroundView?.addSubview(blackMask!)
            }
            
            backgroundView?.isHidden = false
            lastScreenShotView?.removeFromSuperview()
            if let lastScreenShot = screenShots.last {
                lastScreenShotView = UIImageView(image: lastScreenShot)
                backgroundView?.insertSubview(lastScreenShotView!, belowSubview: blackMask!)
            }
            
        case .ended:
            
            if touchPoint.x - startTouch.x > screenWidth * 0.3 {
                UIView.animate(withDuration: 0.25, animations: { [weak self] in
                    self?.moveView(with: self!.screenWidth)
                }) { [weak self] (finished) in
                    _ = self?.popViewController(animated: false)
                    var frame = self!.topView!.frame
                    frame.origin.x = 0
                    UIApplication.shared.keyWindow!.rootViewController!.view.frame = frame
                    
                    self?.isMoving = false
                    self?.backgroundView?.isHidden = true
                }
            } else {
                UIView.animate(withDuration: 0.25, animations: { [weak self] in
                    self?.moveView(with: 0)
                }) { [weak self] (finished) in
                    self?.isMoving = false
                    self?.backgroundView?.isHidden = true
                }
            }
 
            return
            
        case .cancelled:
            
            UIView.animate(withDuration: 0.25, animations: { [weak self] in
                self?.moveView(with: 0)
            }) { [weak self] (finished) in
                self?.isMoving = false
                self?.backgroundView?.isHidden = true
            }
            
            return
            
        default:
            break
        }
        
        if isMoving {
            moveView(with: touchPoint.x - startTouch.x)
        }
        
    }
    
    @objc
    fileprivate func back() {
        _ = popViewController(animated: true)
    }
    
    // MARK: -
    fileprivate func capture() -> UIImage? {
        
        if let topView = topView {
            UIGraphicsBeginImageContextWithOptions(topView.bounds.size, topView.isOpaque, UIScreen.main.scale)
            topView.drawHierarchy(in: topView.bounds, afterScreenUpdates: true)
            topView.layer.contents = nil
            let img = UIGraphicsGetImageFromCurrentImageContext()
            UIGraphicsEndImageContext()

            return img
        }
        
        return nil
    }
    
    fileprivate func barButtonItem(with image: UIImage, higlightedImage: UIImage, target: Any, action: Selector) -> UIBarButtonItem {
        let btn = UIButton(type: .custom)
        
        let normal = image
        let higlighted = higlightedImage
        btn.setImage(normal, for: .normal)
        btn.setImage(higlighted, for: .highlighted)
        btn.bounds = CGRect(x: 0, y: 0, width: 44, height: 44)
        btn.contentHorizontalAlignment = .left
        btn.addTarget(self, action: action, for: .touchUpInside)
        return UIBarButtonItem(customView: btn)
    }
    
    fileprivate func moveView(with x: CGFloat) {
        let x = max(0, x > screenWidth ? screenWidth : x)
        
        var frame = topView!.frame
        frame.origin.x = x
        UIApplication.shared.keyWindow!.rootViewController!.view.frame = frame
        
        let alpha = 0.3 - (x / screenWidth * 0.3)
        let scale = (x / screenWidth * 0.05) + 0.95
        
        switch popAnimationType {
        case .liner:
            if let last = lastScreenShotView {
                last.transform = CGAffineTransform(translationX: x * 0.5 - last.center.x, y: 0)
            }
            break
        case .scale:
            if let last = lastScreenShotView {
                last.transform = CGAffineTransform(scaleX: scale, y: scale)
            }
            break
        default:
            break
        }
        
        blackMask?.alpha = alpha
    }
}

extension HFNavigationController: UIGestureRecognizerDelegate {
    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        return viewControllers.count > 1 && canIneractive ? true : false
    }
}
